specs/_features/eip7251/beacon-chain.md

```diff
@@ -0,0 +1,763 @@
+        if bls.Verify(pubkey, signing_root, signature):
+            state.validators.append(get_validator_from_deposit(pubkey, withdrawal_credentials))
+            state.balances.append(0)
+            state.previous_epoch_participation.append(ParticipationFlags(0b0000_0000))
+            state.current_epoch_participation.append(ParticipationFlags(0b0000_0000))
+            state.inactivity_scores.append(uint64(0))
+            index = len(state.validators) - 1
+            state.pending_balance_deposits.append(PendingBalanceDeposit(index=index, amount=amount))
+    else:
+        index = ValidatorIndex(validator_pubkeys.index(pubkey))
+        state.pending_balance_deposits.append(PendingBalanceDeposit(index=index, amount=amount))
+```
+
+###### Updated `get_validator_from_deposit`
+
+```python
+def get_validator_from_deposit(pubkey: BLSPubkey, withdrawal_credentials: Bytes32) -> Validator:
+    return Validator(
+        pubkey=pubkey,
+        withdrawal_credentials=withdrawal_credentials,
+        activation_eligibility_epoch=FAR_FUTURE_EPOCH,
+        activation_epoch=FAR_FUTURE_EPOCH,
+        exit_epoch=FAR_FUTURE_EPOCH,
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+        effective_balance=0, # [Modified in EIP7251]
+    )
+```
+
+##### Withdrawals 
+
+###### New `process_execution_layer_withdraw_request`
+
+```python
+def process_execution_layer_withdraw_request(
+        state: BeaconState,
+        execution_layer_withdraw_request: ExecutionLayerWithdrawRequest
+    ) -> None:
+    validator_pubkeys = [v.pubkey for v in state.validators]
+    validator_index = ValidatorIndex(validator_pubkeys.index(execution_layer_withdraw_request.validator_pubkey))
+    validator = state.validators[validator_index]
+    amount = execution_layer_withdraw_request.amount
+
+    # Same conditions as in EIP7002 https://github.com/ethereum/consensus-specs/pull/3349/files#diff-7a6e2ba480d22d8bd035bd88ca91358456caf9d7c2d48a74e1e900fe63d5c4f8R223
+    # Verify withdrawal credentials
+    is_execution_address = has_eth1_withdrawal_credential(validator) or has_compounding_withdrawal_credential(validator)
+    is_correct_source_address = validator.withdrawal_credentials[12:] == execution_layer_withdraw_request.source_address
+    if not (is_execution_address and is_correct_source_address):
+        return
+    # Verify the validator is active
+    if not is_active_validator(validator, get_current_epoch(state)):
+        return
+    # Verify exit has not been initiated, and slashed
+    if validator.exit_epoch != FAR_FUTURE_EPOCH:
+        return
+    # Verify the validator has been active long enough
+    if get_current_epoch(state) < validator.activation_epoch + SHARD_COMMITTEE_PERIOD:
+        return
+
+    # New condition: only allow partial withdrawals witb compounding withdrawal credentials
+    is_full_exit_request = amount == 0
+    if not (is_full_exit_request or has_compounding_withdrawal_credential(validator)):
```

Comment:
```suggestion
    if not (amount == FULL_EXIT_REQUEST_AMOUNT or has_compounding_withdrawal_credential(validator)):
```

Suggesting to introduce a constant for the amount denoting a full exit

Reply:
would we ever want to set it to something other than 0?